### PR TITLE
Code Linting with Pylint

### LINKED ISSUE #108: Use flake8 to better verify PEP 8 compliance
This is complementary to using *Black* (#102). There is also overlap with #66. 

### About flake8
[flake8](https://flake8.pycqa.org/en/latest/index.html) checks code compliance to [PEP 8](https://www.python.org/dev/peps/pep-0008/). This is not a duplication to our usage of *Black*: *Black* formats code automatically, but there are code issues that *Black* can not automatically solve, e.g. [wildcard imports](https://docs.quantifiedcode.com/python-anti-patterns/maintainability/from_module_import_all_used.html). flake8 captures these issues.

Sidenote: Trying out flake8 locally seems to raise mostly the following issues:
- line-length (~75%)
- star imports (~10%)
- unused imports (~10%)

There is also a flake8 plugin which I encountered together with flake8 (in the [*Black* repo](https://github.com/psf/black/blob/master/.flake8) actually): [flake8-bugbear](https://github.com/PyCQA/flake8-bugbear). The provided [list of warnings](https://github.com/PyCQA/flake8-bugbear#list-of-warnings) gives me the impression that it overall provides useful feedback.

#### My personal opinion
Cons:
- Yet another tool
- Additional set up requirements for developers
- Another larger code migration

Pros:
- Better code quality through better PEP 8 compliance
- Less work for third-party / beginner PR reviews since feedback on these code quality aspects is given automatically

### Concrete Proposal
- using both flake8 and flake8-bugbear
- adding flake8 to tox, and thus to TravisCI
  (alternatively we could probably also use GitHub Actions)
- adding a pre-commit hook to support that workflow, similar to what I did for `black`

If we chose to do this, then again we will need some form of code *migration* through a larger commit.
I believe that this should best be done right after merging (#102).

### PR DESCRIPTION
**Pull request by nathanaelbosch**
Closes #108 (and relates to #111).

__Current state of the PR__
I created configs, added a github action, and started fixing many errors. However I did not fix *all* issues, and often chose to add exceptions. I think that the best way would be to discuss these exceptions, fix those that should *absolutely* go into this PR, and then open specific issues about individual points.

__Commentary for each of the exceptions:__
- `invalid-name`: We will probably want to keep this, since we often want to use *mathy* notation
- `no-else-return`, `no-else-raise`: Not so clear to me what the benefit of fixing this is
- `fixme`: Can be resolved through the separate issue #69
- `too-many-arguments`: Appears accross modules. Good code style improvement, but not crucial. We could also raise the number of arguments from  to 6 or 7 to resolve most warnigns, and to put the focus towards the cases with 12 (`distribution.py`) and 10 (`linearsolvers.py`) arguements.
- `missing-module-docstring`, `missing-function-docstring`: Could be considered as part of #95, or we can also create a new issue. Although I have the impression that noone will do this at once.
- `abstract-method`: Needs to be fixed in `probnum.linalg` and `probnum.prob`
- `unused-argument`: Many of these are cases where `**kwargs` gets passed around. I had the impression that this is often the "standard" in our code so I did not go ahead and fix it.
- `too-many-branches`: Only affects `linearsolves.py` and `matrixbased.py`. I'd just open an issue for this specific point.
- `arguments-differ`: Multiple cases in `linalg`, but also appearsx in `prob`, `quad` and `diffeq`
- `redefined-builtin`: In `quad` and `linalg`; Fixing this will change some function signatures
- `too-few-public-methods`: Mostly `quad`, but also `diffeq.ODESolver`; Suggests considering removing these classes
- `too-many-locals`: Seems like a good tool to improve code, but is also non-trivial to fix
- `no-self-use`: Suggests making functions instead of methods
- `too-many-lines`: Suggests refactoring `matrixbased`
- `broad-except`: Only appears in `normal.py`. Should be easier to fix for someone familiar with that code.
- `abstract-class-instantiated`: Only appears in `diffeq` due to Nico's code structure; I think it would be best if he has a look.
- `protected-access`: Something we probably want, but it's also not trivial or fast to fix
- `duplicate-code`: Only appears in `filtsmooth`
- `too-complex`: McCabe complexity checker; Complains for `matrixbased`
- `too-many-instance-attributes`: Indicated that `matrixbased.py` and `distribution` might need some refactoring
- `too-many-statements`: Again in `matrixbased`
- `attribute-defined-outside-init`: Mostly in `matrixbased`. One appearance in `bayesquadrature.py`, but that code seems empty anyways.

__More exceptions for tests:__
`duplicate-code`, `line-too-long`, `missing-class-docstring`, `unnecessary-pass`. I did not fix them becasue it did not feel too important for now, but especially the `line-too-long` and `unnecessary-pass` should be easy to fix quite mechanically. Also `unused-variable`, but in tests we often want to compute some value without actually using it afterwards, so that seems reasonable.

__Next steps__
I think it would be good to merge a first version as soon as possible, and just open issues for the parts that we want to have fixed.
But it would be good to know if you think that we should still resolve one of the current exceptions in this PR.
If we decide to merge, I'll open individual issues for the remaining points.

Also, if you have any suggestion on how to better include pylint into our CI let me know!

### CODE REVIEW

**Comment by JonathanWenger** [comment]
In general, I think all these issues raised should be fixed (obviously) but none in this PR specifically. For some there are already issues (e.g. for `linalg`: #51) and some might be fixed by larger changes in progress at the moment (e.g. #83, #131). In summary I agree this PR should be merged and some of the points above moved into (existing) issues.

File: src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py [hunk 1]
```diff
@@ -88,10 +88,10 @@ def __call__(self, t, smoothed=True):
             else:
                 return pred_rv
 
-        elif t > self.locations[-1]:
-            if smoothed:
-                warn("`smoothed=True` is ignored for extrapolation.")
-            return self._predict_to_loc(t)
+        # else: t > self.locations[-1]:
+        if smoothed:
```
Comment (JonathanWenger) [inline src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py:92]: Should this be
```
if smoothed and t > self.locations[-1]:
```
to retain the same behaviour? Clearly no tests are failing so I'm not sure, but according to codecov this is also not covered.

File: src/probnum/filtsmooth/statespace/continuous/linearsdemodel.py [hunk 2]
```diff
@@ -25,15 +25,15 @@ class LinearSDEModel(continuousmodel.ContinuousModel):
 
     Parameters
     ----------
-    driftmatrixfct : callable, signature=(t, \*\*kwargs)
+    driftmatrixfct : callable, signature=(t, **kwargs)
```
Comment (JonathanWenger) [inline src/probnum/filtsmooth/statespace/continuous/linearsdemodel.py:28]: These changes cause warnings in Sphinx (see https://travis-ci.org/github/probabilistic-numerics/probnum/jobs/718931687):
```bash
/home/travis/build/probabilistic-numerics/probnum/src/probnum/filtsmooth/__init__.py:docstring of probnum.filtsmooth.LinearSDEModel:9: WARNING: Inline strong start-string without end-string.

/home/travis/build/probabilistic-numerics/probnum/src/probnum/filtsmooth/__init__.py:docstring of probnum.filtsmooth.LinearSDEModel:13: WARNING: Inline strong start-string without end-string.

/home/travis/build/probabilistic-numerics/probnum/src/probnum/filtsmooth/__init__.py:docstring of probnum.filtsmooth.LinearSDEModel:17: WARNING: Inline strong start-string without end-string.

/home/travis/build/probabilistic-numerics/probnum/src/probnum/filtsmooth/gaussfiltsmooth/unscentedtransform.py:docstring of probnum.filtsmooth.UnscentedTransform.propagate:8: WARNING: Inline strong start-string without end-string.
```

 and likely do not produce the docs output we are looking for.

File: src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py (hunk 1 above)
Comment (nathanaelbosch) [reply src/probnum/filtsmooth/gaussfiltsmooth/kalmanposterior.py:92]: The "if" statements before check for `t < self.locations[0]` and `self.locations[0] < t < self.locations[-1]`, so the check is not actually necessary. I just wrote it that way as first to make the code more readable, and thought now that the comment achieves the same thing.

File: src/probnum/filtsmooth/statespace/continuous/linearsdemodel.py (hunk 2 above)
Comment (nathanaelbosch) [reply src/probnum/filtsmooth/statespace/continuous/linearsdemodel.py:28]: My bad, thought I had checked those things but apparently not for all of those string changes! I'll look into fixing that such that the builds look as they should, while pylint does not complain.

Comment (nathanaelbosch) [reply src/probnum/filtsmooth/statespace/continuous/linearsdemodel.py]: Solved this in nathanaelbosch:probnum/ pylint @6e847a5

**Comment by nathanaelbosch** [comment]
A few things that should be done as a part of this PR:
- perform the pylint check on a "per-module" basis. This allows us to check for things that are currently violated in one of the modules and make sure that we do not introduce new issues.
- Open refactoring issues for the individual modules. Fixing these errors on a per-module basis is easier. Also, we can raise the standards after fixing each individual error, for each module separately.

**Review by JonathanWenger — APPROVED** [review]

**Pull request merged**